README.md
# Energenie

Python module to control the [Energenie](https://energenie4u.co.uk/) add-on board for the [Raspberry Pi](http://www.raspberrypi.org/) used for remotely turning power sockets on and off.

## Installation

On Raspberry Pi, install the `energenie` module in `pip`.

### Python 3:

```bash
sudo apt-get install python3-pip
sudo pip-3.2 install energenie
```

### Python 2:

```bash
sudo apt-get install python-pip
sudo pip install energenie
```

## Usage

```python
from energenie import switch_on, switch_off
from time import sleep

# turn all plug sockets on and off
switch_on()
switch_off()

# turn a plug socket on and off by number
switch_on(1)
switch_off(1)

switch_on(3)
switch_off(3)

# turn some plug sockets on, then turn them off after 10 seconds
switch_on(1)
switch_on(4)
sleep(10)
switch_off(1)
switch_off(4)
```

See full documentation at [pythonhosted.org/energenie](https://pythonhosted.org/energenie/)

## Contributors

- [Ben Nuttall](https://github.com/bennuttall) (project maintainer)
- [Amy Mather](https://github.com/minigirlgeek)
- [Gordon Hollingworth](https://github.com/ghollingworth)

## Package information

- [PyPi](https://pypi.python.org/pypi/energenie) energenie

## Licence

[OSI Approved BSD Licence](LICENCE.txt)


debian/changelog
energenie (1.0.1-1) UNRELEASED; urgency=medium

  * Initial release.

 -- Ben Nuttall <ben@raspberrypi.org>  Mon, 29 Dec 2014 14:56:59 +0000


tests/test_basic_functionality.py
from mock import patch, MagicMock, call
from nose.tools import assert_greater_equal, assert_equal

# Energenie I/O pins in bit 4, 3, 2 and 1 order
_IO_PINS = [27, 23, 22, 17]

# Energenie control pins
_ON_OFF = 24
_ENABLE = 25

# We need to explicitly create a mock for RPio.GPIO since we want to have the
# mock used at import time. Create a mocked RPi.GPIO module which we can use to
# test energenie without a) having to have the hardware installed and/or b)
# having to run on the pi.
_RPi_mock = MagicMock()
_RPi_mock.GPIO = MagicMock()

# This is the dict which should be patched into sys.modules to make imports
# work.
_mocked_modules = {
    'RPi': _RPi_mock,
    'RPi.GPIO': _RPi_mock.GPIO,
}

_GPIO_PATCHER = patch.dict('sys.modules', _mocked_modules)

@_GPIO_PATCHER
def test_energenie_initial_setup():
    """Tests that energenie initialises the module correctly at import time."""
    import energenie
    GPIO = _RPi_mock.GPIO

    GPIO.setmode.assert_called_once_with(GPIO.BCM)
    GPIO.setwarnings.assert_called_once_with(False)

    # Check pins are set up for output
    for pin in _IO_PINS + [_ON_OFF, _ENABLE]:
        GPIO.setup.assert_any_call(pin, GPIO.OUT)

    # Check all pins are low
    for pin in _IO_PINS + [_ON_OFF, _ENABLE]:
        GPIO.output.assert_any_call(pin, False)

def _assert_enabled_toggled():
    # The enable pin should've been toggled high and *then* low
    _RPi_mock.GPIO.output.assert_has_calls([
        call(_ENABLE, True),
        call(_ENABLE, False),
    ])

def _assert_one_on(socket_num):
    # Assert that a single socket was switched on
    code = [0b1111, 0b1110, 0b1101, 0b1100][socket_num - 1]
    for bit in range(4):
        value = True if code & (1<<bit) != 0 else False
        _RPi_mock.GPIO.output.assert_any_call(_IO_PINS[3-bit], value)

def _assert_one_off(socket_num):
    # Assert that a single socket was switched off
    code = [0b0111, 0b0110, 0b0101, 0b0100][socket_num - 1]
    for bit in range(4):
        value = True if code & (1<<bit) != 0 else False
        _RPi_mock.GPIO.output.assert_any_call(_IO_PINS[3-bit], value)

@_GPIO_PATCHER
def test_switch_all_on():
    from energenie import switch_on

    GPIO = _RPi_mock.GPIO
    GPIO.output.reset_mock()
    assert_equal(GPIO.output.call_count, 0)

    switch_on()

    # We should've set at least six pins
    assert_greater_equal(GPIO.output.call_count, 6)

    GPIO.output.assert_any_call(_IO_PINS[0], True)
    GPIO.output.assert_any_call(_IO_PINS[1], False)
    GPIO.output.assert_any_call(_IO_PINS[2], True)
    GPIO.output.assert_any_call(_IO_PINS[3], True)

    _assert_enabled_toggled()

@_GPIO_PATCHER
def test_switch_all_off():
    from energenie import switch_off

    GPIO = _RPi_mock.GPIO
    GPIO.output.reset_mock()
    assert_equal(GPIO.output.call_count, 0)

    switch_off()

    # We should've set at least six pins
    assert_greater_equal(GPIO.output.call_count, 6)

    GPIO.output.assert_any_call(_IO_PINS[0], False)
    GPIO.output.assert_any_call(_IO_PINS[1], False)
    GPIO.output.assert_any_call(_IO_PINS[2], True)
    GPIO.output.assert_any_call(_IO_PINS[3], True)

    _assert_enabled_toggled()

@_GPIO_PATCHER
def test_switch_on_individual():
    from energenie import switch_on

    # Test switching on each socket in turn
    for socket in range(1,5):
        print("Testing socket {0}".format(socket))
        _RPi_mock.GPIO.output.reset_mock()
        switch_on(socket)
        print(_RPi_mock.GPIO.output.mock_calls)
        _assert_one_on(socket)
        _assert_enabled_toggled()

@_GPIO_PATCHER
def test_switch_off_individual():
    from energenie import switch_off

    # Test switching off each socket in turn
    for socket in range(1,5):
        print("Testing socket {0}".format(socket))
        _RPi_mock.GPIO.output.reset_mock()
        switch_off(socket)
        print(_RPi_mock.GPIO.output.mock_calls)
        _assert_one_off(socket)
        _assert_enabled_toggled()


energenie/energenie.py
import RPi.GPIO as GPIO
from time import sleep

# The GPIO pins for the Energenie module
BIT1 = 17
BIT2 = 22
BIT3 = 23
BIT4 = 27

ON_OFF_KEY = 24
ENABLE = 25

GPIO.setmode(GPIO.BCM)
GPIO.setwarnings(False)

GPIO.setup(BIT1, GPIO.OUT)
GPIO.setup(BIT2, GPIO.OUT)
GPIO.setup(BIT3, GPIO.OUT)
GPIO.setup(BIT4, GPIO.OUT)

GPIO.setup(ON_OFF_KEY, GPIO.OUT)
GPIO.setup(ENABLE, GPIO.OUT)

GPIO.output(ON_OFF_KEY, False)
GPIO.output(ENABLE, False)

GPIO.output(BIT1, False)
GPIO.output(BIT2, False)
GPIO.output(BIT3, False)
GPIO.output(BIT4, False)

# Codes for switching on and off the sockets
#       all     1       2       3       4
ON = ['1011', '1111', '1110', '1101', '1100']
OFF = ['0011', '0111', '0110', '0101', '0100']


def change_plug_state(socket, on_or_off):
    state = on_or_off[socket][3] == '1'
    GPIO.output(BIT1, state)
    state = on_or_off[socket][2] == '1'
    GPIO.output(BIT2, state)
    state = on_or_off[socket][1] == '1'
    GPIO.output(BIT3, state)
    state = on_or_off[socket][0] == '1'
    GPIO.output(BIT4, state)
    sleep(0.1)
    GPIO.output(ENABLE, True)
    sleep(0.25)
    GPIO.output(ENABLE, False)


def switch_on(socket=0):
    change_plug_state(socket, ON)


def switch_off(socket=0):
    change_plug_state(socket, OFF)
